src/neo4j_graphrag/generation/graphrag.py

```diff
@@ -124,18 +127,25 @@ def search(
             )
         except ValidationError as e:
             raise SearchValidationError(e.errors())
-        query_text = validated_data.query_text
+        query = self.build_query(validated_data.query_text, chat_history)
         retriever_result: RetrieverResult = self.retriever.search(
-            query_text=query_text, **validated_data.retriever_config
+            query_text=query, **validated_data.retriever_config
         )
         context = "\n".join(item.content for item in retriever_result.items)
         prompt = self.prompt_template.format(
             query_text=query_text, context=context, examples=validated_data.examples
         )
         logger.debug(f"RAG: retriever_result={retriever_result}")
         logger.debug(f"RAG: prompt={prompt}")
-        answer = self.llm.invoke(prompt)
+        answer = self.llm.invoke(prompt, chat_history)
         result: dict[str, Any] = {"answer": answer.content}
         if return_context:
             result["retriever_result"] = retriever_result
         return RagResultModel(**result)
+    
+    def build_query(self, query_text: str, chat_history: list[dict[str, str]]) -> str:
+        if chat_history:
+            summarization_prompt = ChatSummaryTemplate().format(chat_history=chat_history)
+            summary = self.llm.invoke(summarization_prompt).content
```

Comment:
Just realised that we have to reset the `system_instruction` on the llm for this to work. And this will be a problem for `VertexAI` because `system_instruction` is set directly on the GenerativeModel object.

Reply:
Would that work if we instantiate a client for each call to `invoke` (for VertexAI I mean)?

Reply:
Something like that. To be continued next week 🙂

Reply:
WDYT about [this](https://github.com/neo4j/neo4j-graphrag-python/pull/234/commits/5bb56f6130dff46ced90e693e7bdeaa7b5339173)? It's getting quite convoluted, I don't know if there's a simpler way 🤔

Reply:
It makes sense, I like the idea of having the system instructions with the prompt, they are quite tied together. 

Do we still need the system instruction in the init of the LLMInterface if we have it in the invoke method?

Reply:
> Do we still need the system instruction in the init of the LLMInterface if we have it in the invoke method?

Technically, no. But then it would be required to send the system instruction on each invokation, which would be kind of contradictory to it's purpose (to set it once, as a frame for the conversation), at least the way I understand it.

Reply:
Yeah, but having said that, the system instruction is just another message, that gets inserted at the beginning of the message list, so it could be treated in the same way. Also, I think the `llm` object could be used for different purposes. For instance, consider GraphRAG with a Text2Cypher retriever: the same LLM object can serve as text2cypher generator and answer generator. In that case, we won't use the system instructions in the init.

Reply:
But how are you thinking it should work in the case of chat conversation? We want to be able to set the system instruction once before the conversation starts, and avoid having to send it along with each user message, right?